src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -571,21 +583,21 @@ private static byte[] UrlEncodeNonAscii(byte[] bytes, int offset, int count)
                 return value;
             }
 
-            string? schemeAndAuthority;
-            string? path;
-            string? queryAndFragment;
+            Range schemeAndAuthority;
+            Range path;
+            Range queryAndFragment;
 
             if (!UriUtil.TrySplitUriForPathEncode(value, out schemeAndAuthority, out path, out queryAndFragment))
             {
                 // If the value is not a valid url, we treat it as a relative url.
                 // We don't need to extract query string from the url since UrlPathEncode()
                 // does not encode query string.
-                schemeAndAuthority = null;
-                path = value;
-                queryAndFragment = null;
+                schemeAndAuthority = default(Range);
+                path = Range.All;
+                queryAndFragment = default(Range);
```

Comment:
```suggestion
                return UrlPathEncodeImpl(value);
```

Otherwise you would end up allocating a substring for `value.AsSpan(path).ToString()` + a substring for the `string.Concat`.